### PR TITLE
Implement wait_until_available and reconnects (part of RFC1004)

### PR DESCRIPTION
**Pull request by tailhook**
This is should be equivalent to edgedb/edgedb-js#82, and requires slightly updated edgedb/edgedb#2061 to include tags in `errors.txt`.

To do:
* [x] synchronous implementation

### CODE REVIEW

File: edgedb/errors/tags.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+    'Tag',
+    'SHOULD_RECONNECT',
+]
+
+
+class Tag(object):
+    pass
+
+
+SHOULD_RECONNECT = Tag()
```
Comment (1st1) [inline edgedb/errors/tags.py:11]: Let's do better than this:

```

class Tag(object):
    def __repr__(self):
        return f'<edgedb.{self.name}>'

SHOULD_RECONNECT = Tag('SHOULD_RECONNECT')
```

File: edgedb/errors/_base.py [hunk 2]
```diff
@@ -79,11 +79,15 @@ class EdgeDBErrorMeta(Meta):
 class EdgeDBError(Exception, metaclass=EdgeDBErrorMeta):
 
     _code = None
+    tags = frozenset(())
```
Comment (1st1) [inline edgedb/errors/_base.py:82]:
```suggestion
    tags = frozenset()
```

File: edgedb/errors/__init__.py [hunk 3]
```diff
@@ -362,6 +381,24 @@ class ClientConnectionError(ClientError):
...
 
+class ConnectionFailedError(ClientConnectionError):
+    _code = 0x_FF_01_01_00
+
+
+class ConnectionFailedTemporarilyError(ConnectionFailedError):
+    _code = 0x_FF_01_01_01
+
+
+    tags = frozenset({SHOULD_RECONNECT})
```
Comment (1st1) [inline edgedb/errors/__init__.py:392]: There's one extra line before the `tags = ` line.

File: edgedb/base_con.py [hunk 4]
```diff
@@ -101,6 +102,7 @@ def add_log_listener(
             **connection**: a Connection the callback is registered with;
             **message**: the `edgedb.EdgeDBMessage` message.
         """
+        # TODO(tailhook)
```
Comment (1st1) [inline edgedb/base_con.py:105]: Can you clarify this and the following TODOs?

File: edgedb/asyncio_con.py [hunk 5]
```diff
@@ -32,41 +35,160 @@
...
+            else:
+                tr, pr = await loop.create_connection(factory, *addr)
+        except socket.gaierror as e:
+            # All name resolution errors are considered temporary
+            raise errors.ConnectionFailedTemporarilyError(str(e)) from e
+        except OSError as e:
+            # TODO(tailhook) figure out error
+            message = str(e)
+            if e.errno is None:
+                errnos = extract_errno(message)
```
Comment (1st1) [inline edgedb/asyncio_con.py:133]: Why not `e.errno`?

File: edgedb/asyncio_con.py [hunk 6]
```diff
@@ -32,41 +35,160 @@
...
+
+    def __init__(self, codecs_registry, query_cache):
+        self._codecs_registry = codecs_registry
+        self._query_cache = query_cache
+
+        # TODO(tailhook) temporary, this is not needed when we implement
+        # transaction part of RFC1004
+        self._top_xact = None
 
-    def __instancecheck__(cls, instance):
```
Comment (1st1) [inline edgedb/asyncio_con.py:44]: Why was this deleted?

Comment (1st1) [reply edgedb/asyncio_con.py:44]: Because it's now implemented via `abstract.AsyncIOExecutor`, right?

Comment (tailhook) [reply edgedb/asyncio_con.py:44]: No. Because `PoolConnection` is a subclass of `AsyncIOConnection` anyway. And we use two hooks (most notably `_reconnect`) to make it working for pools instead of using Proxy pattern and metaclasses.

File: edgedb/errors/tags.py (hunk 1 above)
Comment (tailhook) [reply edgedb/errors/tags.py]: Fixed.

File: edgedb/base_con.py (hunk 4 above)
Comment (tailhook) [reply edgedb/base_con.py]: No need for `is_closed` check. Removed the check and TODO.

File: edgedb/asyncio_con.py (hunk 5 above)
Comment (tailhook) [reply edgedb/asyncio_con.py]: Because it's absent in "Multiple exceptions" case. I've added a docstring in `_extract_errno` text

**Comment by tailhook** [comment]
Note: we don't put method name like `AsyncIOConnection.fetch_one` in interface errors. But:
1. This is quite rare error: using connection after `with acquire()` block ended
2. This is now visible in traceback (previous implementation raised error in wrapper in metaclass, so exact method called wasn't visible in a traceback).

File: edgedb/asyncio_con.py [hunk 7]
```diff
@@ -32,41 +35,160 @@
...
+
+        while True:
+            for addr in addrs:
+                try:
+                    await asyncio.wait_for(
+                        self._connect_addr(loop, addr, params),
+                        config.connect_timeout,
+                    )
+                except asyncio.TimeoutError as e:
+                    if iteration == 1 or time.monotonic() < max_time:
```
Comment (1st1) [inline edgedb/asyncio_con.py:94]: Re `iteration == 1 or ...`: so if `wait_until_available` is 0 we'd still give it two tries?

Comment (tailhook) [reply edgedb/asyncio_con.py:94]: Iterations start from `1`. When `wait_until_available` is zero we give a one try for each of the addresses specified.

**Review by 1st1 — APPROVED** [review]
